### PR TITLE
Dynamic container junit5

### PR DESCRIPTION
**Pull request by vchaitanya**
AbstractFlocessor flows() provides all chained transactions grouped together in the flattened list. 
After the change, these chained transactions are grouped together under a dynamic container

### CODE REVIEW

File: assert/assert-junit5/src/main/java/com/mastercard/test/flow/assrt/junit5/Flocessor.java [hunk 1]
```diff
@@ -53,33 +60,72 @@ public Flocessor( String title, Model model ) {
...
-						// not strictly required to record the error outcome in the history, as it
-						// does not inform the processing of later flows. That may change in the future
-						// though, so for now we're going to live with the mutation testing complaint
-						history.recordResult( flow, Result.ERROR );
-						throw e;
-					}
-				} ) );
+		List<DynamicNode> nodes = new ArrayList<>();
+		List<Flow> currentChain = new ArrayList<>();
+		boolean inChain = false;
```
[the remark this answers is not in the document]
Comment (therealryan) [reply assert/assert-junit5/src/main/java/com/mastercard/test/flow/assrt/junit5/Flocessor.java:65]: It is important to maintain the order - we've done a bunch of complex work to make sure that inter-flow dependencies are being honoured

[the remark this answers is not in the document]
Comment (vchaitanya) [reply assert/assert-junit5/src/main/java/com/mastercard/test/flow/assrt/junit5/Flocessor.java]: Updated the code logic.

**Comment by therealryan** [comment]
I've taken the liberty of writing a test that dumps out the container/test structure:
```java

/**
 * A simple sequence of flows with no chains
 */
@Test
void simple() {
	expectNodes( model( null, null, null, null, null ),
			"test : 0 []",
			"test : 1 []",
			"test : 2 []",
			"test : 3 []",
			"test : 4 []" );
}

/**
 * A single chain of a single flow
 */
@Test
void link() {
	expectNodes( model( null, "a", null ),
			"test : 0 []",
			"container : Chain: 1 [chain:a]",
			"  test : 1 [chain:a]",
			"test : 2 []" );
}

/**
 * Consecutive single-flow chains
 */
@Test
void links() {
	expectNodes( model( "a", "b", "c" ),
			"container : Chain: 0 [chain:a]",
			"  test : 0 [chain:a]",
			"container : Chain: 1 [chain:b]",
			"  test : 1 [chain:b]",
			"container : Chain: 2 [chain:c]",
			"  test : 2 [chain:c]" );
}

/**
 * A single multi-flow chain
 */
@Test
void chain() {
	// in the middle
	expectNodes( model( null, "a", "a", "a", null ),
			"test : 0 []",
			"container : Chain: 1 [chain:a]",
			"  test : 1 [chain:a]",
			"  test : 2 [chain:a]",
			"  test : 3 [chain:a]",
			"test : 4 []" );

	// at the start
	expectNodes( model( "a", "a", "a", null ),
			"container : Chain: 0 [chain:a]",
			"  test : 0 [chain:a]",
			"  test : 1 [chain:a]",
			"  test : 2 [chain:a]",
			"test : 3 []" );

	// at the end
	expectNodes( model( null, "a", "a", "a" ),
			"test : 0 []",
			"container : Chain: 1 [chain:a]",
			"  test : 1 [chain:a]",
			"  test : 2 [chain:a]",
			"  test : 3 [chain:a]" );
}

/**
 * Multiple multi-flow chains
 */
@Test
void chains() {
	expectNodes( model( "a", "a", null, "b", "b", "c" ),
			"container : Chain: 0 [chain:a]",
			"  test : 0 [chain:a]",
			"  test : 1 [chain:a]",
			"test : 2 []",
			"container : Chain: 3 [chain:b]",
			"  test : 3 [chain:b]",
			"  test : 4 [chain:b]",
			"container : Chain: 5 [chain:c]",
			"  test : 5 [chain:c]" );
}

private static Model model( String... chains ) {
	List<Flow> flows = new ArrayList<>();
	for( int i = 0; i < chains.length; i++ ) {
		int idx = i;
		Flow flow = Creator
				.build( f -> f.meta( data -> data
						.description( String.valueOf( idx ) )
						.tags( Tags.add( Optional.ofNullable( chains[idx] )
								.map( v -> Chain.PREFIX + v )
								.orElse( "" ) ) ) ) );
		flows.add( flow );
	}
	Model model = mock( Model.class );
	when( model.flows( anySet(), anySet() ) )
			.thenReturn( flows.stream() );

	return model;
}

private static void expectNodes( Model model, String... expected ) {
	Flocessor flocessor = new Flocessor( "", model );
	List<String> actual = new ArrayList<>();
	flocessor.tests()
			.forEach( node -> stringify( node, "", actual ) );
	assertEquals(
			copypasta( Stream.of( expected ) ),
			copypasta( actual.stream() ) );
}

private static void stringify( DynamicNode node, String prefix, List<String> lines ) {
	if( node instanceof DynamicTest ) {
		DynamicTest test = (DynamicTest) node;
		lines.add( prefix + "test : " + test.getDisplayName() );
	}
	else if( node instanceof DynamicContainer ) {
		DynamicContainer container = (DynamicContainer) node;
		lines.add( prefix + "container : " + container.getDisplayName() );
		container.getChildren()
				.forEach( child -> stringify( child, prefix + "  ", lines ) );
	}
	else {
		throw new IllegalStateException( "unexpected node " + node.getClass() );
	}
}

/**
 * @param content Some strings
 * @return A string that can be trivially copy/pasted into java source
 */
private static String copypasta( Stream<String> content ) {
	return content
			.map( s -> s.replaceAll( "\r", "" ) )
			.flatMap( s -> Stream.of( s.split( "\n" ) ) )
			.map( s -> s.replaceAll( "\"", "'" ) )
			.collect( Collectors.joining( "\",\n\"", "\"", "\"" ) );
}
```

The flow structure looks good, but I think we should change the display name of the `DynamicContainer`s - they should just be the `chain:foobar` tag rather than the name of the first flow in the chain.

File: assert/assert-junit5/src/main/java/com/mastercard/test/flow/assrt/junit5/Flocessor.java [hunk 2]
```diff
@@ -53,33 +60,82 @@ public Flocessor( String title, Model model ) {
...
+
+	private DynamicContainer createDynamicContainer( List<Flow> chain ) {
+		List<DynamicTest> tests = chain.stream()
+				.map( flow -> dynamicTest(
+						flow.meta().id(),
+						testSource( flow ),
+						() -> processFlow( flow ) ) )
+				.collect( Collectors.toList() );
+
+		return DynamicContainer.dynamicContainer( "Chain: " + chain.get( 0 ).meta().id(), tests );
```
[the remark this answers is not in the document]
Comment (vchaitanya) [reply assert/assert-junit5/src/main/java/com/mastercard/test/flow/assrt/junit5/Flocessor.java]: Updated

**Comment by therealryan** [comment]
Looks great!

**Pull request merged**